source/server/admin/listeners_handler.cc

```diff
@@ -29,14 +29,28 @@ Http::Code ListenersHandler::handlerDrainListeners(Http::ResponseHeaderMap&,
     return Http::Code::BadRequest;
   }
   if (graceful) {
-    // Ignore calls to /drain_listeners?graceful if the drain sequence has
-    // already started.
+    auto direction = Network::DrainDirection::All;
+    if (stop_listeners_type == ListenerManager::StopListenersType::InboundOnly) {
+      direction = Network::DrainDirection::InboundOnly;
+    }
+    // If this returns true, it means:
+    // 1. we are already draining
+    // 2. That drain includes the direction we're being asked to drain
+    // We should just return a 200
+    auto duplicate_drain = server_.drainManager().drainClose(direction);
+    if (duplicate_drain) {
+      response.add("OK\n");
+      return Http::Code::OK;
+    }
+    // This means either we aren't draining or we still have to do some work
+    // (e.g. we were draining inbound only but now we're being asked to drain all)
     if (!server_.drainManager().draining()) {
```

Comment:
Remove this check; the server may be draining but we still need to drain a different direction